File: app/src/main/java/io/github/anonymous/pepper_realtime/FaceRecognitionService.java
package io.github.anonymous.pepper_realtime;

import android.graphics.Bitmap;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;

@SuppressWarnings({"SpellCheckingInspection", "ConstantConditions"})
public class FaceRecognitionService {

    private static final String TAG = "FaceRecognitionService";

    public interface FaceCallback {
        void onResult(List<FaceInfo> faces);
        void onError(Exception e);
    }

    public static class RateLimitException extends Exception {
        public final long retryAfterMs;
        public RateLimitException(long retryAfterMs) {
            super("Azure Face API rate limited");
            this.retryAfterMs = retryAfterMs;
        }
    }

    public static class FaceInfo {
        public int left;
        public int top;
        public int width;
        public int height;
        public Double yawDeg; // headPose.yaw in degrees
        public Double pitchDeg;
        public Double rollDeg;
        public String glassesType; // e.g., NoGlasses, ReadingGlasses, Sunglasses
        public Boolean isMasked;
        public String imageQuality; // low|medium|high
        public Double blurValue; // 0..1
        public String exposureLevel; // underExposure|goodExposure|overExposure
    }

    private final OkHttpClient http;

    public FaceRecognitionService() {
        this.http = OptimizedHttpClientManager.getInstance().getApiClient();
    }

    public boolean isConfigured() {
        final String endpoint = BuildConfig.AZURE_FACE_ENDPOINT;
        final String key = BuildConfig.AZURE_FACE_API_KEY;
        return endpoint != null && endpoint.length() > 0 && key != null && key.length() > 0;
    }

    public void detectFacesAsync(Bitmap image, FaceCallback callback) {
        if (!isConfigured() || image == null) {
            if (callback != null) callback.onResult(Collections.emptyList());
            return;
        }
        OptimizedThreadManager.getInstance().executeNetwork(() -> {
            try {
                // Downscale to reduce bandwidth/load
                Bitmap scaled = image;
                try {
                    final int maxW = 640;
                    if (image.getWidth() > maxW) {
                        int nh = (int) (image.getHeight() * (maxW / (double) image.getWidth()));
                        scaled = Bitmap.createScaledBitmap(image, maxW, nh, true);
                    }
                } catch (Exception ignore) { }

                // Prepare JPEG bytes
                ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
                scaled.compress(Bitmap.CompressFormat.JPEG, 80, outputStream);
                byte[] imageBytes = outputStream.toByteArray();

                // Build request
                String base = BuildConfig.AZURE_FACE_ENDPOINT;
                if (base.endsWith("/")) base = base.substring(0, base.length() - 1);
                String url = base + "/face/v1.0/detect" +
                        "?returnFaceId=false" +
                        "&detectionModel=detection_03" +
                        "&recognitionModel=recognition_04" +
                        "&returnFaceAttributes=headPose,glasses,mask,blur,exposure,qualityForRecognition";

                Request req = new Request.Builder()
                        .url(url)
                        .addHeader("Ocp-Apim-Subscription-Key", BuildConfig.AZURE_FACE_API_KEY)
                        .addHeader("Content-Type", "application/octet-stream")
                        .post(RequestBody.create(imageBytes, MediaType.get("application/octet-stream")))
                        .build();

                try (Response resp = http.newCall(req).execute()) {
                    ResponseBody rb = resp.body();
                    if (!resp.isSuccessful() || rb == null) {
                        if (resp.code() == 429) {
                            String ra = resp.header("Retry-After");
                            long retryMs = 15000L;
                            try { if (ra != null) retryMs = Long.parseLong(ra.trim()) * 1000L; } catch (Exception ignore) {}
                            Log.w(TAG, "Azure detect failed: " + resp + ", retryAfterMs=" + retryMs);
                            if (callback != null) callback.onError(new RateLimitException(retryMs));
                        } else {
                            Log.w(TAG, "Azure detect failed: " + resp);
                            if (callback != null) callback.onResult(Collections.emptyList());
                        }
                        return;
                    }
                    String body = rb.string();
                    JSONArray arr = new JSONArray(body);
                    List<FaceInfo> results = new ArrayList<>();
                    for (int i = 0; i < arr.length(); i++) {
                        JSONObject obj = arr.getJSONObject(i);
                        FaceInfo info = new FaceInfo();
                        try {
                            JSONObject rect = obj.optJSONObject("faceRectangle");
                            if (rect != null) {
                                info.left = rect.optInt("left");
                                info.top = rect.optInt("top");
                                info.width = rect.optInt("width");
                                info.height = rect.optInt("height");
                            }
                            JSONObject attrs = obj.optJSONObject("faceAttributes");
                            if (attrs != null) {
                                JSONObject headPose = attrs.optJSONObject("headPose");
                                if (headPose != null) {
                                    info.yawDeg = headPose.optDouble("yaw");
                                    info.pitchDeg = headPose.optDouble("pitch");
                                    info.rollDeg = headPose.optDouble("roll");
                                }
                                String glasses = attrs.optString("glasses", "");
                                info.glassesType = glasses.isEmpty() ? null : glasses;
                                JSONObject mask = attrs.optJSONObject("mask");
                                if (mask != null) {
                                    if (mask.has("noseAndMouthCovered")) {
                                        info.isMasked = mask.optBoolean("noseAndMouthCovered");
                                    } else {
                                        String maskType = mask.optString("type", "");
                                        info.isMasked = !maskType.isEmpty() && !"noMask".equalsIgnoreCase(maskType);
                                    }
                                }
                                JSONObject blur = attrs.optJSONObject("blur");
                                if (blur != null) {
                                    info.blurValue = blur.optDouble("value");
                                }
                                JSONObject exposure = attrs.optJSONObject("exposure");
                                if (exposure != null) {
                                    String level = exposure.optString("exposureLevel", "");
                                    info.exposureLevel = level.isEmpty() ? null : level;
                                }
                                String quality = attrs.optString("qualityForRecognition", "");
                                if (!quality.isEmpty()) info.imageQuality = quality;
                            }
                        } catch (Exception ignore) { }
                        results.add(info);
                    }
                    Log.i(TAG, "Azure detect OK, faces=" + results.size());
                    if (callback != null) callback.onResult(results);
                }
            } catch (Exception e) {
                Log.e(TAG, "Face detect REST failed", e);
                if (callback != null) callback.onError(e);
            }
        });
    }
}

File: app/src/main/java/io/github/anonymous/pepper_realtime/service/FaceRecognitionService.kt
package io.github.anonymous.pepper_realtime.service

import android.graphics.Bitmap
import android.util.Log
import io.github.anonymous.pepper_realtime.BuildConfig
import io.github.anonymous.pepper_realtime.di.IoDispatcher
import io.github.anonymous.pepper_realtime.network.HttpClientManager
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import java.io.ByteArrayOutputStream
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Service for Azure Face API integration.
 * Provides face detection with attributes like head pose, glasses, mask, etc.
 */
@Suppress("SpellCheckingInspection")
@Singleton
class FaceRecognitionService @Inject constructor() {
    // For DI injection (optional - not currently used)
    constructor(
        @Suppress("UNUSED_PARAMETER") httpClientManager: HttpClientManager,
        @Suppress("UNUSED_PARAMETER") @IoDispatcher ioDispatcher: CoroutineDispatcher
    ) : this()
    companion object {
        private const val TAG = "FaceRecognitionService"
    }

    /**
     * Legacy callback interface for Java interop
     */
    interface FaceCallback {
        fun onResult(faces: List<FaceInfo>)
        fun onError(e: Exception)
    }

    /**
     * Exception thrown when Azure rate limits the request
     */
    class RateLimitException(val retryAfterMs: Long) : Exception("Azure Face API rate limited")

    /**
     * Face detection result with attributes
     */
    data class FaceInfo(
        var left: Int = 0,
        var top: Int = 0,
        var width: Int = 0,
        var height: Int = 0,
        var yawDeg: Double? = null,
        var pitchDeg: Double? = null,
        var rollDeg: Double? = null,
        var glassesType: String? = null,
        var isMasked: Boolean? = null,
        var imageQuality: String? = null,
        var blurValue: Double? = null,
        var exposureLevel: String? = null
    )

    private val http by lazy { HttpClientManager.getInstance().getApiClient() }

    /**
     * Check if Azure Face API is configured
     */
    fun isConfigured(): Boolean {
        val endpoint = BuildConfig.AZURE_FACE_ENDPOINT
        val key = BuildConfig.AZURE_FACE_API_KEY
        return !endpoint.isNullOrEmpty() && !key.isNullOrEmpty()
    }

    /**
     * Detect faces in an image using Azure Face API.
     * This is a suspend function - preferred for new Kotlin code.
     *
     * @param image Bitmap to analyze
     * @return List of detected faces with attributes
     * @throws RateLimitException if Azure rate limits the request
     */
    suspend fun detectFaces(image: Bitmap?): List<FaceInfo> = withContext(Dispatchers.IO) {
        if (!isConfigured() || image == null) {
            return@withContext emptyList()
        }

        try {
            // Downscale to reduce bandwidth/load
            val scaled = try {
                val maxW = 640
                if (image.width > maxW) {
                    val nh = (image.height * (maxW / image.width.toDouble())).toInt()
                    Bitmap.createScaledBitmap(image, maxW, nh, true)
                } else {
                    image
                }
            } catch (e: Exception) {
                image
            }

            // Prepare JPEG bytes
            val outputStream = ByteArrayOutputStream()
            scaled.compress(Bitmap.CompressFormat.JPEG, 80, outputStream)
            val imageBytes = outputStream.toByteArray()

            // Build request URL
            var base = BuildConfig.AZURE_FACE_ENDPOINT
            if (base.endsWith("/")) base = base.dropLast(1)
            val url = "$base/face/v1.0/detect" +
                    "?returnFaceId=false" +
                    "&detectionModel=detection_03" +
                    "&recognitionModel=recognition_04" +
                    "&returnFaceAttributes=headPose,glasses,mask,blur,exposure,qualityForRecognition"

            val request = Request.Builder()
                .url(url)
                .addHeader("Ocp-Apim-Subscription-Key", BuildConfig.AZURE_FACE_API_KEY)
                .addHeader("Content-Type", "application/octet-stream")
                .post(imageBytes.toRequestBody("application/octet-stream".toMediaType()))
                .build()

            http.newCall(request).execute().use { response ->
                val body = response.body

                if (!response.isSuccessful || body == null) {
                    if (response.code == 429) {
                        val retryAfter = response.header("Retry-After")
                        val retryMs = try {
                            retryAfter?.trim()?.toLong()?.times(1000L) ?: 15000L
                        } catch (e: Exception) {
                            15000L
                        }
                        Log.w(TAG, "Azure detect failed: $response, retryAfterMs=$retryMs")
                        throw RateLimitException(retryMs)
                    } else {
                        Log.w(TAG, "Azure detect failed: $response")
                        return@withContext emptyList()
                    }
                }

                val responseBody = body.string()
                val arr = JSONArray(responseBody)
                val results = mutableListOf<FaceInfo>()

                for (i in 0 until arr.length()) {
                    val obj = arr.getJSONObject(i)
                    val info = FaceInfo()

                    try {
                        obj.optJSONObject("faceRectangle")?.let { rect ->
                            info.left = rect.optInt("left")
                            info.top = rect.optInt("top")
                            info.width = rect.optInt("width")
                            info.height = rect.optInt("height")
                        }

                        obj.optJSONObject("faceAttributes")?.let { attrs ->
                            attrs.optJSONObject("headPose")?.let { headPose ->
                                info.yawDeg = headPose.optDouble("yaw")
                                info.pitchDeg = headPose.optDouble("pitch")
                                info.rollDeg = headPose.optDouble("roll")
                            }

                            val glasses = attrs.optString("glasses", "")
                            info.glassesType = glasses.ifEmpty { null }

                            attrs.optJSONObject("mask")?.let { mask ->
                                info.isMasked = if (mask.has("noseAndMouthCovered")) {
                                    mask.optBoolean("noseAndMouthCovered")
                                } else {
                                    val maskType = mask.optString("type", "")
                                    maskType.isNotEmpty() && !maskType.equals("noMask", ignoreCase = true)
                                }
                            }

                            attrs.optJSONObject("blur")?.let { blur ->
                                info.blurValue = blur.optDouble("value")
                            }

                            attrs.optJSONObject("exposure")?.let { exposure ->
                                val level = exposure.optString("exposureLevel", "")
                                info.exposureLevel = level.ifEmpty { null }
                            }

                            val quality = attrs.optString("qualityForRecognition", "")
                            if (quality.isNotEmpty()) info.imageQuality = quality
                        }
                    } catch (e: Exception) {
                        // Ignore parsing errors for individual faces
                    }

                    results.add(info)
                }

                Log.i(TAG, "Azure detect OK, faces=${results.size}")
                results
            }
        } catch (e: RateLimitException) {
            throw e
        } catch (e: Exception) {
            Log.e(TAG, "Face detect REST failed", e)
            throw e
        }
    }

    /**
     * Legacy async method for Java interop.
     * New Kotlin code should use the suspend function detectFaces() instead.
     */
    @Deprecated("Use suspend fun detectFaces() for new Kotlin code")
    fun detectFacesAsync(image: Bitmap?, callback: FaceCallback?) {
        if (!isConfigured() || image == null) {
            callback?.onResult(emptyList())
            return
        }

        // Use a simple thread for legacy compatibility
        Thread({
            try {
                // This is a blocking call - not ideal but maintains compatibility
                val httpClient = HttpClientManager.getInstance().getApiClient()
                
                // Downscale
                val scaled = try {
                    val maxW = 640
                    if (image.width > maxW) {
                        val nh = (image.height * (maxW / image.width.toDouble())).toInt()
                        Bitmap.createScaledBitmap(image, maxW, nh, true)
                    } else {
                        image
                    }
                } catch (e: Exception) {
                    image
                }

                val outputStream = ByteArrayOutputStream()
                scaled.compress(Bitmap.CompressFormat.JPEG, 80, outputStream)
                val imageBytes = outputStream.toByteArray()

                var base = BuildConfig.AZURE_FACE_ENDPOINT
                if (base.endsWith("/")) base = base.dropLast(1)
                val url = "$base/face/v1.0/detect" +
                        "?returnFaceId=false" +
                        "&detectionModel=detection_03" +
                        "&recognitionModel=recognition_04" +
                        "&returnFaceAttributes=headPose,glasses,mask,blur,exposure,qualityForRecognition"

                val request = Request.Builder()
                    .url(url)
                    .addHeader("Ocp-Apim-Subscription-Key", BuildConfig.AZURE_FACE_API_KEY)
                    .addHeader("Content-Type", "application/octet-stream")
                    .post(imageBytes.toRequestBody("application/octet-stream".toMediaType()))
                    .build()

                httpClient.newCall(request).execute().use { response ->
                    val body = response.body

                    if (!response.isSuccessful || body == null) {
                        if (response.code == 429) {
                            val retryAfter = response.header("Retry-After")
                            val retryMs = try {
                                retryAfter?.trim()?.toLong()?.times(1000L) ?: 15000L
                            } catch (parseEx: Exception) {
                                15000L
                            }
                            callback?.onError(RateLimitException(retryMs))
                        } else {
                            callback?.onResult(emptyList())
                        }
                        return@use
                    }

                    val responseBody = body.string()
                    val arr = JSONArray(responseBody as String)
                    val results = mutableListOf<FaceInfo>()

                    for (i in 0 until arr.length()) {
                        val obj = arr.getJSONObject(i)
                        val info = FaceInfo()

                        try {
                            obj.optJSONObject("faceRectangle")?.let { rect: org.json.JSONObject ->
                                info.left = rect.optInt("left")
                                info.top = rect.optInt("top")
                                info.width = rect.optInt("width")
                                info.height = rect.optInt("height")
                            }

                            obj.optJSONObject("faceAttributes")?.let { attrs: org.json.JSONObject ->
                                attrs.optJSONObject("headPose")?.let { headPose: org.json.JSONObject ->
                                    info.yawDeg = headPose.optDouble("yaw")
                                    info.pitchDeg = headPose.optDouble("pitch")
                                    info.rollDeg = headPose.optDouble("roll")
                                }

                                val glasses = attrs.optString("glasses", "")
                                info.glassesType = glasses.ifEmpty { null }

                                attrs.optJSONObject("mask")?.let { mask: org.json.JSONObject ->
                                    info.isMasked = if (mask.has("noseAndMouthCovered")) {
                                        mask.optBoolean("noseAndMouthCovered")
                                    } else {
                                        val maskType = mask.optString("type", "")
                                        maskType.isNotEmpty() && !maskType.equals("noMask", ignoreCase = true)
                                    }
                                }

                                attrs.optJSONObject("blur")?.let { blur: org.json.JSONObject ->
                                    info.blurValue = blur.optDouble("value")
                                }

                                attrs.optJSONObject("exposure")?.let { exposure: org.json.JSONObject ->
                                    val level = exposure.optString("exposureLevel", "")
                                    info.exposureLevel = level.ifEmpty { null }
                                }

                                val quality = attrs.optString("qualityForRecognition", "")
                                if (quality.isNotEmpty()) info.imageQuality = quality
                            }
                        } catch (parseEx: Exception) {
                            // Ignore parsing errors for individual faces
                        }

                        results.add(info)
                    }

                    Log.i(TAG, "Azure detect OK, faces=${results.size}")
                    callback?.onResult(results)
                }
            } catch (e: RateLimitException) {
                callback?.onError(e)
            } catch (e: Exception) {
                Log.e(TAG, "Face detect REST failed", e)
                callback?.onError(e)
            }
        }, "face-recognition-thread").start()
    }
}


